system-tests-new/tiny-project/test_tiny_project.py

```diff
@@ -1,56 +1,217 @@
+        assert rebuild_hashes == clean_hashes
+        assert rebuild_csvs == clean_csvs
+
+    def test_incremental_build_no_mod(self, configs):
+        # test incremental fortran build, no module change
+        grab_config, build_config = configs
+        grab_config.run()
+
+        # clean build
+        clean_timestamps, clean_hashes, clean_csvs = self.build(build_config)
+
+        # modify the fortran module source without changing the module interface
+        mod_source = build_config.source_root / 'src/my_mod.F90'
+        lines = open(mod_source, 'rt').readlines()
+        with open(mod_source, 'wt') as out:
+            for line in lines:
+                out.write(line)
+                # duplicate the print line
+                if 'PRINT' in line:
+                    out.write(line)
+
+        # rebuild
+        rebuild_timestamps, rebuild_hashes, _ = self.build(build_config)
+
+        # check only the object file has changed, not the mod file
+        build_output = build_config.project_workspace / BUILD_OUTPUT
+
+        changed_items = set(rebuild_timestamps.items()) - set(clean_timestamps.items())
+        changed_paths = [i[0] for i in changed_items]
+        assert build_output / 'src/my_mod.o' in changed_paths
+        assert build_output / 'my_mod.mod' not in changed_paths
+
+    def test_incremental_build_mod(self, configs):
+        # test incremental fortran build with module change
+        grab_config, build_config = configs
+        grab_config.run()
+
+        # clean build
+        clean_timestamps, clean_hashes, clean_csvs = self.build(build_config)
+
+        # modify the fortran module source, changing the module interface
+        mod_source = build_config.source_root / 'src/my_mod.F90'
+        lines = open(mod_source, 'rt').readlines()
+        with open(mod_source, 'wt') as out:
+            for line in lines:
+                out.write(line)
+                # add an extra subroutine
+                if 'END SUBROUTINE' in line:
+                    out.write("""
+                        SUBROUTINE added_func ()
+
+                            INTEGER :: bar = 2
+                            PRINT *, bar
 
-        # record the file timestamps
+                        END SUBROUTINE added_func
+                    """)
 
-    def no_change_rebuild(self):
-        pass
+        # rebuild
+        rebuild_timestamps, rebuild_hashes, _ = self.build(build_config)
 
-    def incremental_build(self):
-        # modify the source
-        pass
+        # check both the object file and mod file have changed
+        build_output = build_config.project_workspace / BUILD_OUTPUT
 
-        # check only the right stuff has changed
+        changed_items = set(rebuild_timestamps.items()) - set(clean_timestamps.items())
+        changed_paths = [i[0] for i in changed_items]
+        assert build_output / 'src/my_mod.o' in changed_paths
+        assert build_output / 'my_mod.mod' in changed_paths
```

Comment:
I think this is mostly testing the compiler's ability to only regenerate the module file if the API has changed. It's not clear to me that the actual dependency is being tested. i.e. when the "my_mod" API changes then "my_prog" which uses that API should be rebuilt too.

Given that the API has been extended and "my_prog" is clear that it only uses `mod_func` it might be reasonable not to rebuild but I don't think Fab is able to make that determination.

Reply:
:+1: That's now tested with the improved thoroughness from previous comments. The timestamps, hashes and csvs are all checked.